hail/python/hail/fs/router_fs.py

```diff
@@ -232,13 +237,92 @@ async def size():
         return _stat_result(
             *await asyncio.gather(fle.is_dir(), size(), fle.url()))
 
-    def ls(self, path: str, _max_simultaneous_files: int = 50) -> List[StatResult]:
-        async def _ls():
-            async with OnlineBoundedGather2(asyncio.Semaphore(_max_simultaneous_files)) as pool:
-                tasks = [pool.call(self._fle_to_dict, fle)
-                         async for fle in await self.afs.listfiles(path)]
-                return list(await asyncio.gather(*tasks))
-        return async_to_blocking(_ls())
+    def ls(self,
+           path: str,
+           *,
+           error_when_file_and_directory: bool = True,
+           _max_simultaneous_files: int = 50) -> List[StatResult]:
+        return async_to_blocking(self._async_ls(
+            path,
+            error_when_file_and_directory=error_when_file_and_directory,
+            _max_simultaneous_files=_max_simultaneous_files))
+
+    async def _async_ls(self,
+                        path: str,
+                        *,
+                        error_when_file_and_directory: bool = True,
+                        _max_simultaneous_files: int = 50) -> List[StatResult]:
+        async def ls_no_glob(path) -> List[StatResult]:
+            return await self._ls_no_glob(path,
+                                          error_when_file_and_directory=error_when_file_and_directory,
+                                          _max_simultaneous_files=_max_simultaneous_files)
+        url = self.afs.parse_url(path)
+        if any(glob.escape(bucket_part) != bucket_part
+               for bucket_part in url.bucket_parts):
+            raise ValueError(f'glob pattern only allowed in path (e.g. not in bucket): {path}')
+
+        blobpath = url.path
+        if isinstance(url, LocalAsyncFSURL) and blobpath[0] != '/':
+            blobpath = './' + blobpath
+
+        components = blobpath.split('/')
+        assert len(components) > 0
+
+        glob_components = []
+        running_prefix = []
+
+        for component in components:
+            _raise_for_incomplete_glob_group(component, path)
+            if glob.escape(component) == component:
+                running_prefix.append(component)
+            else:
+                glob_components.append((running_prefix, component))
+                running_prefix = []
+
+        suffix_components: List[str] = running_prefix
+        first_prefix = '/'.join([url.scheme +':', '', *url.bucket_parts])
+        cumulative_prefixes = [first_prefix]
+
+        for intervening_components, single_component_glob_pattern in glob_components:
```

Comment:
Hard for me to tell what intervening means. I feel like `no_pattern_components` is more straightforward.

Reply:
preceeding is fine